### ISSUE TITLE
doc: clarify repeating timer semantics

### ISSUE THREAD

**Issue by skeggse** [libuv/libuv#181]
After reading the docs on how repeating timers work, I'd like further clarification as to whether a repeating timer will "drift." If you're using javascript, `setInterval` "drifts" over time, in that v8 just schedules a new timer after the previous one finishes. That means that if the callback takes 15 milliseconds to do its job, the next timer will happen 15 milliseconds late, and the error compounds. Do libuv timers function in a similar way? I would imagine not, given that `uv_timer_set_repeat` will "take effect **when possible**", as the next timer may have already been scheduled.

**Comment by txdv** [comment]
It is not drifting, since it uses the system clocktime before every loop iteration.

**Comment by skeggse** [comment]
Awesome. Would love to see this reflected in the docs.

**Comment by bnoordhuis** [comment]
I suppose it depends on how you define 'drift'.  The logic is as follows:

1. Libuv updates its idea of 'now'.
2. For each expired timer:
3. Rearm the timer
4. Invoke the callback
5. Goto 2.

That means when:

1. Two repeating 50 ms timers expire at the same "tick" of the event loop, and
2. If the first timer callback takes 14 ms to complete, and
3. The second timer callback takes 1 ms to complete, then
4. The timer callbacks will run again 35 ms after step 3.

(Or 4b: in the case of a time slice overrun, the callbacks fire as soon as possible.)

Additionally, timer callbacks may call `uv_update_time()` to update libuv's idea of now.

Pull requests that clarify the documentation on this point are certainly welcome. :-)

**Comment by saghul** [comment]
Fixed by #520.

**Comment by dkrikun** [comment]
I am not able to observe this behavior. I set a timer with a repeat value of 1000 ms, which blocks for 200 ms. Then I print current time in the beginning of the callback and it shows that the callback is invoked each 1200 ms.
How it could be possibly explained?

**Comment by bnoordhuis** [comment]
@dkrikun Can you file a new issue and include a test case?
